### PR TITLE
feat: remove sampling from the topos-network

### PR DESCRIPTION
**Pull request by gruberb**
Removing of the `sampling.rs` file, adjusting the `TCE-Transport` parameters, and adjusting double-echo to have no sampling but taking the whole network (deterministic) as an input.

Fixes [TP-587](https://linear.app/toposware/issue/TP-587/as-topos-we-want-to-remove-the-samples-so-that-we-can-proceed-to-the)

# Description

We want to merge the TCE Network and the topos subnet, therefore the idea of nodes sampling a subset of nodes for their `echo | ready | delivery ` phase is not needed anymore. We want all nodes to communicate and know each other, instead of just a subset of them.

Therefore:

* We remove the `echo_sample_size`, `ready_sample_size` and `delivery_sample_size` everywhere (including the ENV variables)
* Instead of passing the sample sizes down to different functions, we create a `visible_peers` variable and pass it down to the needed functions and calculations
* `visible_peers` are all the nodes in the new TCE network.

## Additions and Changes

Currently, we do not use the `ReliableBroadcastParams` anywhere but in the tests. Therefore [I added a `allow_dead_code` parameter](https://github.com/topos-network/topos/pull/221/commits/b72859ad04ff6c7e2afe4b0f0d366421ab5c346c#diff-d8fcc1dab4cf0411809470b7b5afb39b020a236da45f20c98e2fe260593cedf8R108-R109). 

## Breaking changes

* No more `TCE_ECHO_SAMPLE_SIZE`, `TCE_READY_SAMPLE_SIZE` and `TCE_DELIVERY_SAMPLE_SIZE` environment variable ([see here](https://github.com/topos-network/topos/pull/221/files#diff-12790b9d233a58892d6e076ed36aa6e550ffb8960d3df38d483e9f7229626eccL67-L69))

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-tce-broadcast/src/lib.rs [hunk 1]
```diff
@@ -121,7 +121,7 @@ impl ReliableBroadcastClient {
     /// Aggregate is spawned as new task.
     // #[instrument(name = "ReliableBroadcastClient", skip_all)]
     pub fn new(
-        config: ReliableBroadcastConfig,
+        _config: ReliableBroadcastConfig,
```
Comment (gruberb) [inline crates/topos-tce-broadcast/src/lib.rs:124]: Turns out, we just needed the `sample` parameters when creating the `ReliableBroadcastClient`. I am not sure about that yet, seems a bit strange.

File: crates/topos-tce-broadcast/src/sampler/mod.rs [hunk 2]
```diff
@@ -355,86 +374,62 @@ impl Sampler {
...
-                    error!("Unable to send event {:?}", error);
-                }
-            }
-            Err(e) => {
-                error!("Failed to create the sample for the Echo Subscriptions: {e:?}");
-            }
+        for peer in &self.visible_peers {
+            info!("Adding the Peer {peer} to the pending Echo Subscriptions");
+            self.pending_subscriptions.echo.insert(*peer);
+        }
```
Comment (atanmarko) [inline crates/topos-tce-broadcast/src/sampler/mod.rs:385]:
```suggestion
     self.pending_subscriptions.echo = self.visible_peers.iter().cloned().collect();
```

Comment (atanmarko) [reply crates/topos-tce-broadcast/src/sampler/mod.rs:385]: Also on other similar places.

File: crates/topos-tce-broadcast/src/tests/mod.rs [hunk 3]
```diff
@@ -2,7 +2,7 @@ use crate::double_echo::*;
 use crate::mem_store::TceMemStore;
 use crate::sampler::SubscribersUpdate;
 use crate::*;
-use rand::seq::IteratorRandom;
+// use rand::seq::IteratorRandom;
```
Comment (atanmarko) [inline crates/topos-tce-broadcast/src/tests/mod.rs:5]: Remove

File: crates/topos-tce/src/app_context.rs [hunk 4]
```diff
@@ -326,7 +326,7 @@ impl AppContext {
                 })
                 .into();
                 let command_sender = self.tce_cli.get_sampler_channel();
-                // Sending echo subscribe message to send to a number of remote peers
+                // Sending echo subscribe message to send all remote peers
```
Comment (atanmarko) [inline crates/topos-tce/src/app_context.rs:329]:
```suggestion
                // Sending echo subscribe message to all remote peers
```

File: crates/topos-tce-broadcast/src/double_echo/mod.rs [hunk 5]
```diff
@@ -435,20 +446,23 @@ impl DoubleEcho {
             }
 
             let echo_missing = self
-                .params
-                .echo_sample_size
+                .subscribers
+                .get_subscribers()
```
Comment (Freyskeyd) [inline crates/topos-tce-broadcast/src/double_echo/mod.rs:450]: You're comparing two things that doesn't fit together.
`Subscribers` are people interested into receiving your messages (Echo/Ready), `Subscriptions` are peers that you want to receive message from. At some point the `DoubleEcho` will not have `subscribers` anymore.
You need to base the calculation on `network_size - current_subscription`, the current subscriptions are reduced when receiving a message upon reaching the threshold.

Example:

```
---- state ----
current node: peer1
all threshold: 2
network_size: 10 - 1 (peer1 is you)
Echo Subscribers: [ peer2, peer3, peer4, peer5, ..., peer10 ]
Echo Subscriptions: [ peer2, peer3, peer4, peer5, ..., peer10 ]
----

-> RECV: Echo from peer2
    -> Remove peer2 from Echo Subscriptions

Echo Subscriptions: [ peer3, peer4, peer5, ..., peer10 ]

Threshold ? network_size - len(Echo Subscriptions) -> 9 - 8 = 1 : NO

----

-> RECV: Echo from peer4
    -> Remove peer4 from Echo Subscriptions

Echo Subscriptions: [ peer3, peer5, ..., peer10 ]

Threshold ? network_size - len(Echo Subscriptions) -> 9 - 7 = 2 : YES

```

File: crates/topos-tce-broadcast/src/tests/mod.rs [hunk 6]
```diff
@@ -38,29 +33,17 @@ fn small_config() -> TceParams {
 fn medium_config() -> TceParams {
     TceParams {
         nb_peers: 50,
-        echo_subscribers_sample_size: 5,
```
Comment (Freyskeyd) [inline crates/topos-tce-broadcast/src/tests/mod.rs:41]: Maybe we need to update the threshold regarding the new size, from 20 to 50. The values can be found in Notion.

**Review by Freyskeyd — CHANGES_REQUESTED** [review]

File: crates/topos-tce-broadcast/src/sampler/mod.rs [hunk 7]
```diff
@@ -100,14 +105,17 @@ impl SubscriptionsView {
 }
 
 pub struct Sampler {
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (gruberb) [reply crates/topos-tce-broadcast/src/sampler/mod.rs:108]: We use the `params` in the tests, but not in "production". I was contemplating with a different `Sampler` struct and add a feature flag, or `allow(dead_code)`. Not sore what would be better here.

File: crates/topos-tce-broadcast/src/double_echo/mod.rs (hunk 5 above)
Comment (gruberb) [reply crates/topos-tce-broadcast/src/double_echo/mod.rs:450]: Thank you so much for the explanation, now it finally clicked. Makes sense 👍

File: crates/topos-tce-broadcast/src/sampler/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-broadcast/src/sampler/mod.rs:108]: Let's keep the `params` for now

File: crates/topos-tce-broadcast/src/sampler/mod.rs [hunk 8]
```diff
@@ -71,8 +70,13 @@ pub enum SubscribersUpdate {
     RemoveReadySubscribers(HashSet<PeerId>),
 }
 
+// Same set, merge
+// Need to seperate echo and ready (echo removes it from echo set, ready removes it from ready set)
+// HashSet is all participants, once I receive echo | ready | delivery, I remove it to get to the threshold
+// Maybe structure for keeping track of different counters
```
[the remark this answers is not in the document]
Comment (gruberb) [reply crates/topos-tce-broadcast/src/sampler/mod.rs:74]: Added a ticket!

File: crates/topos-tce-broadcast/src/sampler/mod.rs (hunk 2 above)
Comment (atanmarko) [reply crates/topos-tce-broadcast/src/sampler/mod.rs:385]: All the peers from the visible peers are already printed above with debug? `info` is too verbose I think

File: crates/topos-tce-broadcast/src/sampler/mod.rs [hunk 9]
```diff
@@ -661,6 +643,7 @@ mod tests {
         }
 
         sampler.peers_changed(peers).await;
+        sampler.pending_subscriptions.network_size = nb_peers as usize;
```
Comment (Freyskeyd) [inline crates/topos-tce-broadcast/src/sampler/mod.rs:646]: You can put that in the same place as the `self.subscriptions.network_size` one

File: crates/topos-tce-broadcast/src/tests/mod.rs (hunk 6 above)
Comment (Freyskeyd) [reply crates/topos-tce-broadcast/src/tests/mod.rs:41]: @gruberb you still need to update the different threshold regarding the comment above

File: crates/topos-tce/src/app_context.rs [hunk 10]
```diff
@@ -266,11 +266,11 @@ impl AppContext {
             }
 
             ProtocolEvents::Broadcast { certificate_id } => {
-                info!("Broadcasting certificate {}", certificate_id);
+                warn!("Broadcasting certificate {}", certificate_id);
```
Comment (Freyskeyd) [inline crates/topos-tce/src/app_context.rs:269]: Ah those two changes should be revert to info and debug

File: crates/topos/tests/cert_delivery.rs [hunk 11]
```diff
@@ -53,16 +53,16 @@ async fn start_a_cluster() {
...
         .with_test_writer()
         .finish();
-    let _ = ::tracing::subscriber::set_global_default(subscriber);
+    let _ = tracing::subscriber::set_global_default(subscriber);
 
-    let peer_number = 15;
-    let correct_sample = 5;
-    let number_of_certificates_per_subnet = 2;
-    let number_of_subnets = 3;
+    let peer_number = 5;
```
Comment (Freyskeyd) [inline crates/topos/tests/cert_delivery.rs:62]: we can maybe revert to 15 nodes/15 correct sample with the old configuration for number of certificates/subnet

**Review by Freyskeyd — CHANGES_REQUESTED** [review]

**Comment by gruberb** [comment]
I updated the PR @atanmarko @Freyskeyd and fixed your comments!

**Review by Freyskeyd — APPROVED** [review]

**Pull request merged**